### ISSUE TITLE
NULL if malloc (capacity + HEADER_SIZE + 1) for new string is lower than buffer size

### ISSUE THREAD

**Issue by akaufmann** [manastech/crystal#449]
And it only happens if you `malloc` another string

```crystal
str = String.new(0) do |buf| # change 0 to 3 and it works
  appender = buf.appender
  s = "a" * 10
  data = s.cstr
  s.length.times { |i| appender << data[i] }

  count = appender.count
  {count, count}
end

str2 = String.new(0) do |buf| 
  appender = buf.appender
  s = "b" * 10
  data = s.cstr
  s.length.times { |i| appender << data[i] }

  count = appender.count
  {count, count}
end

# [Decode error - output not utf-8]
puts str # NULL, NULL, ...
puts str2 # bbbbbbbbbb
```

**Comment by kostya** [comment]
appender not reallocate string, for speed, use buf << data[i]

**Comment by akaufmann** [comment]
Did you mean `buf[i] = data[i]`? You get the same result.

The properly capacity (for this example) is 3. I don't think this would be the properly capacity which you would set. 
What happens, if `s.length` in the block is dynamic? Which capacity should you use?

**Comment by asterite** [comment]
Hi,

I just added some documentation to the many `String#new` methods. As @kostya said, `String.new(capacity, &block)` allocates a non-resizable buffer with that capacity, so the resulting string must fit in there. Since you are passing 0, that will never happen.

An issue, that we might want to change, is that we don't check that the bytesize you return in that block is less than the capacity given originally. So basically what happens in your case is that you allocate an empty buffer and write to it, possibly accessing invalid memory.

If you need to create a String for which you don't know the maximum length, use `String.build`.

**Comment by akaufmann** [comment]
That all makes sense. But one thing that confuses me. If I remove the  `str2 = String.new ...` and the capacity for `str`  is still 0, it works even if buffer size is 10 or 10000. ` GC.malloc_atomic((capacity + HEADER_SIZE + 1).to_u32)` allocates with capacity 0 only 13 bytes.

**Comment by asterite** [comment]
@akaufmann I debugged your code a little and I found the "problem". If inside the blocks you put `pp buf.address` it will print the address of the buffer allocated for each string. In my case, it's something like this:

```crystal
buf.address = 4364025852
buf.address = 4364025836
```

That shows that `GC.malloc` probably has allocated some memory and it's gives it to you from the end, decrementing the address. So, when you write to str2's buffer you are overwriting str's contents and then the string becomes garbage.

In short, that method is unsafe. We could make it a little safer by verifying that the returned bytesize is less than the capacity, but nothing prevents you from writing through that pointer outside of the allocated bounds.

In conclusion, any code involving pointers in Crystal is unsafe. That, and C bindings, are the two only unsafe things.

**Comment by akaufmann** [comment]
Thanks for your explanation and your help.
Have you ever thought about to teach programming?! ;-) I like the way you do it.
